contents/java/Spring.md

```diff
@@ -0,0 +1,292 @@
+<i>cross-cutting concerns</i></tooltip>. Examples of these concerns include transaction management, authentication, logging, security etc. 
+It does so by adding additional behavior to existing code
+without modifying the code itself, instead separately specifying which code is modified via a "pointcut"
+specification. This allows behaviors that are not central to the business logic (such as logging) to be added to a program
+without cluttering the core code. AOP forms a basis for aspect-oriented software development.
+
+By separating application business logic from system services, Spring Framework supports Aspect Oriented Programming
+and enables cohesive development. It provides the way to dynamically add the
+cross-cutting concern before, after or around the actual logic using simple
+pluggable configurations. 
+
+For example, we wish to log something into the console everytime before we call the method `getEmployeeById`.
+We can then write aspect class annotated with @Aspect annotation and write point-cut expressions to match
+<tooltip content="Joinpoint is a point of execution of the program, such as the execution of a method or the handling of an exception..">
+<i>joint-point</i>
+</tooltip> methods.
+```java
+@Aspect
+public class EmployeeCRUDAspect {
+      
+    @Before("execution(* EmployeeManager.getEmployeeById(..))") // point-cut expressions
+    public void logBefore(JoinPoint joinPoint) {
+        System.out.println("EmployeeCRUDAspect.logBefore() : "
+                + joinPoint.getSignature().getName());
+    }
+    
+    @After("execution(* EmployeeManager.getEmployeeById(..))") // point-cut expressions
+    public void logAfter(JoinPoint joinPoint) {
+        System.out.println("EmployeeCRUDAspect.logAfter() : "
+                + joinPoint.getSignature().getName());
+    }
+}
+```
+    
+And the corresponding `EmployeeManager.java` file could be:
+```java
+@Component
+public class EmployeeManager
+{
+    public EmployeeDTO getEmployeeById(Integer employeeId) {
+        System.out.println("Method getEmployeeById() called");
+        return new EmployeeDTO();
+    }
+}
+```
+In above example, `logBefore()` will be executed before `getEmployeeById()` method because it matches the join-point expression.
+Similarly, `logAfter()` will be executed after `getEmployeeById()`.
+The out put of the above code will be:
+```
+EmployeeCRUDAspect.logBefore() : getEmployeeById
+Method getEmployeeById() called
+EmployeeCRUDAspect.logAfter() : getEmployeeById
+```
+
+The declaration of aspects and advices can also go into XML files so we do not need to write the annotation in our
+source code. This allows easier maintainance of the code as we can add/remove concerns 
+by changing XML configuration file without recompiling source code. The above configuration can be written in XML file as:
+```xml
```

Comment:
Seems that the indentation in this file is inconsistent. Let's keep it to 4 spaces.